mirgecom/steppers.py

```diff
@@ -0,0 +1,158 @@
+    istep = 0
+
+    discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+    nodes = discr.nodes().with_queue(queue)
+    fields = initializer(0, nodes)
+    sdt = get_inviscid_timestep(discr, fields, c=cfl, eos=eos)
+
+    initname = initializer.__class__.__name__
+    eosname = eos.__class__.__name__
+    message = (
+        f"Num {dim}d order-{order} elements: {mesh.nelements}\n"
+        f"Timestep:        {dt}\n"
+        f"Final time:      {t_final}\n"
+        f"Status freq:     {nstepstatus}\n"
+        f"Initialization:  {initname}\n"
+        f"EOS:             {eosname}"
+    )
+    logger.info(message)
+
+    vis = make_visualizer(discr, discr.order + 3 if dim == 2 else discr.order)
+
+    def write_soln(write_status=True):
+        dv = eos(fields)
+        expected_result = initializer(t, nodes)
+        result_resid = fields - expected_result
+        maxerr = [np.max(np.abs(result_resid[i].get())) for i in range(dim + 2)]
+        mindv = [np.min(dvfld.get()) for dvfld in dv]
+        maxdv = [np.max(dvfld.get()) for dvfld in dv]
+
+        if write_status is True:
+            statusmsg = (
+                f"Status: Step({istep}) Time({t})\n"
+                f"------   P({mindv[0]},{maxdv[0]})\n"
+                f"------   T({mindv[1]},{maxdv[1]})\n"
+                f"------   dt,cfl = ({dt},{cfl})\n"
+                f"------   Err({maxerr})"
+            )
+            logger.info(statusmsg)
+
+        io_fields = split_fields(dim, fields)
+        io_fields += eos.split_fields(dim, dv)
+        io_fields.append(("exact_soln", expected_result))
+        io_fields.append(("residual", result_resid))
+        nameform = casename + "-{iorank:04d}-{iostep:06d}.vtu"
+        visfilename = nameform.format(iorank=rank, iostep=istep)
+        vis.write_vtk_file(visfilename, io_fields)
+
+        return maxerr
+
+    def rhs(t, w):
+        return inviscid_operator(discr, w=w, t=t, boundaries=boundaries, eos=eos)
+
+    while t < t_final:
+
+        if constantcfl is True:
+            dt = sdt
+        else:
+            cfl = dt / sdt
+
+        if nstepstatus > 0:
+            if istep % nstepstatus == 0:
+                write_soln()
+
+        fields = rk4_step(fields, t, dt, rhs)
+        t += dt
+        istep += 1
+
+        sdt = get_inviscid_timestep(discr, fields, c=cfl, eos=eos)
+
+    if nstepstatus > 0:
+        logger.info("Writing final dump.")
+        maxerr = max(write_soln(False))
+    else:
+        expected_result = initializer(t, nodes)
+        result_resid = fields - expected_result
+        maxerr = np.max(
+            [
+                np.max(
+                    np.abs(
+                        result_resid[i].get()
+                    )
+                ) for i in range(dim + 2)
+            ]
+        )
+
+    logger.info(f"Max Error: {maxerr}")
+    if maxerr > exittol:
+        raise ValueError("Solution failed to follow expected result.")
+
+    logger.info("Goodbye!")
+    return(maxerr)
```

Comment:
This should return the final state.

Reply:
Even though the caller already "owns" the state?

Reply:
Not sure I understand... the state isn't updated in place.

Reply:
You are right of course. I was doubly confused about (a) python by ref/ by value idiosyncrasies and (b) about how it was implemented even in my own code.   Please have a gander at [lump2.py](https://github.com/illinois-ceesd/mirgecom/blob/mtc/euler/examples/lump2.py) which will be promptly renamed once we converge on what we want the example to have at this point.